sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/Refcounted.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.pubsublite.internal;
+
+import com.google.cloud.pubsublite.internal.wire.SystemExecutors;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A class that safely ensures an object of type T is cleaned up before it is garbage collected. */
+class Refcounted<T extends AutoCloseable> implements Supplier<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Refcounted.class);
+
+  private final T object;
+
+  public static <T extends AutoCloseable> Refcounted<T> of(T object) {
+    return new Refcounted<>(object);
+  }
+
+  @Override
+  public T get() {
+    return object;
+  }
+
+  private Refcounted(T object) {
+    this.object = object;
+  }
+
+  private static class Closer implements Runnable {
+
+    private final AutoCloseable object;
+
+    private Closer(AutoCloseable object) {
+      this.object = object;
+    }
+
+    @Override
+    public void run() {
+      try {
+        object.close();
+      } catch (Exception e) {
+        LOG.warn("Failed to close resource with class: " + object.getClass().getCanonicalName(), e);
+      }
+    }
+  }
+
+  @SuppressWarnings("deprecation")
```

Comment:
What is the deprecation and why is it overridden? Can you add a comment about it please?

Reply:
finalize on Object is deprecated. There are other mechanisms recommended to access the same hooks, but they're wildly more complicated to work with (PhantomReference and friends)

It is unlikely that core java will ever remove finalize() from Object however

Reply:
Ah, I see. Indeed, `PhantomReference` and `Cleaner` and whatnot are the "new" way to do this, but the reasoning there is that they are harder to misuse than `finalize`. So, careful code review and reasoning here suffices to be OK with this use of `finalize`.

Reply:
Done.